### PR TITLE
Added support for CRAM 3.1 using the htscodecs JavaScript.

### PR DESCRIPTION
**Pull request by jkbonfield**
I've been exploring adding my Javascript reference implementations of the CRAM 3.1 htscodecs to cram-js, such to validate they work for real rather than my naive tests.  This PR does that, but it also needs the htscodecs cram-js branch in which I've added an index.js interface to make the functions match the expected prototypes used by existing cram-js.  I've also packaged it up in npm, albeit rather cluelessly probably.   It's my first time ever trying this and I'm not really a JS developer so any advice is welcome.

I also validated that the rans.js in htscodecs is compatible with your own, which it is.  You'll also be pleased to hear your implementation is faster. :-)

Given CRAM 3.1 hasn't been accepted as a final GA4GH standard yet this is perhaps a bit premature, but I'm submitting it as a proof of principle.

Also included is my attempt at a basic cram2sam.js tool.  It's rather hacked together, but it proved useful as a debugging tool so maybe it'd be useful for others.  It may be best to move it to the test directory though.

From commit log:

> Also added a simple cram2sam.js test harness.
> Example usage:
> 
>     node cram2sam.js $HREF38 /var/tmp/novaseq-10m.31.cram chr1 5000000 5100000
> 
> (That's chr1:5000000-5100000 using GRCh38 ref)

### CODE REVIEW

**Comment by cmdcolin** [comment]
>Also included is my attempt at a basic cram2sam.js tool.

Awesome, this is a great way to compare against reference implementations like samtools

**Comment by jkbonfield** [comment]
It doesn't add MD and NM tags.  I thought of adding those to the `decodeSeqCigar` function too as that's where I generate those normally.  This probably hints at missing functionality in the core cram-js library.

Mind you, I confess I'd written this before spotting `getReadBases` existed.  I'd have been kicking myself if it actually did everything I wanted! haha.

File: package.json [hunk 1]
```diff
@@ -44,7 +44,10 @@
   "dependencies": {
     "@babel/runtime-corejs3": "^7.4.5",
     "@gmod/binary-parser": "^1.3.5",
+    "@gmod/indexedfasta": "^1.0.13",
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply package.json]: Done

**Comment by jkbonfield** [comment]
I'm not sure this is ready for merge as I don't really know the policies on NPM and renaming things.  Right now CRAM 3.1 is my own experiment, but it hasn't been formally adopted by GA4GH as we haven't got a second implementation.  (In theory two implementations by myself would count, but it's not ideal and cram-js integration doesn't count either as it is read-only.)

So at some point I expect htscodecs to move from my own personal github account to the samtools organisation.  It probably should have started off there infact, but I have no permissions for that.  This ought to also be true for the npm module.  If that's an issue I can try getting the samtools account admin to set up npm themselves up-front so we can get the final package name used already.

Anyway, for the curious and as a bit of self publicity ;-) here are the fastest of 3 runs each:

```
@ deskpro[work/cram-js]; time node test/cram2sam.js $HREF38 /var/tmp/novaseq-10m.30.cram chr1 5000000 5250000 | md5sum
9b5fd3fbd96fb4067fdc072a0d0f722c  -

real	0m6.999s
user	0m10.390s
sys	0m0.588s

@ deskpro[work/cram-js]; time node test/cram2sam.js $HREF38 /var/tmp/novaseq-10m.31.cram chr1 5000000 5250000 | md5sum
9b5fd3fbd96fb4067fdc072a0d0f722c  -

real	0m6.915s
user	0m10.136s
sys	0m0.506s


@ deskpro[work/cram-js]; time node test/cram2sam.js $HREF38 /var/tmp/novaseq-10m.31s.cram chr1 5000000 5250000 | md5sum
9b5fd3fbd96fb4067fdc072a0d0f722c  -

real	0m7.796s
user	0m12.519s
sys	0m0.604s

-rw-r--r-- 1 jkb team117 223887354 Mar 25 09:55 /var/tmp/novaseq-10m.30.cram
-rw-r--r-- 1 jkb team117 186372268 Mar 25 17:23 /var/tmp/novaseq-10m.31.cram
-rw-r--r-- 1 jkb team117 176651366 Mar 30 14:51 /var/tmp/novaseq-10m.31s.cram
```

30 and 31 refer to CRAM 3.0 and CRAM 3.1 defeault compression modes.  31s is from scramble -V3.1 -Xsmall, which enables fqzcomp and larger block sizes.  It doesn't help a great deal with NovaSeq, but the new rANS with bit-packing and RLE helps hugely so it's a substantial size drop even in normal mode.

On some older data, where fqzcomp qual encoder shines in size (if not speed!), while the newer rANS doesn't make much difference to normal mode:

```
@ deskpro[work/cram-js]; time node test/cram2sam.js $HREF /var/tmp/9827_10m.30.cram 1 10000000 11500000|md5sum
01a1e2bdfeb519224f913968eebd315c  -

real	0m1.931s
user	0m2.388s
sys	0m0.197s


@ deskpro[work/cram-js]; time node test/cram2sam.js $HREF /var/tmp/9827_10m.31.cram 1 10000000 11500000|md5sum
01a1e2bdfeb519224f913968eebd315c  -

real	0m1.983s
user	0m2.360s
sys	0m0.208s


@ deskpro[work/cram-js]; time node test/cram2sam.js $HREF /var/tmp/9827_10m.31small.cram 1 10000000 11500000|md5sum
01a1e2bdfeb519224f913968eebd315c  -

real	0m3.268s
user	0m4.288s
sys	0m0.259s


-rw-r--r-- 1 jkb team117 573186896 Mar 25 09:58 /var/tmp/9827_10m.30.cram
-rw-r--r-- 1 jkb team117 551530728 Mar 25 17:18 /var/tmp/9827_10m.31.cram
-rw-r--r-- 1 jkb team117 390684524 Mar 30 13:56 /var/tmp/9827_10m.31small.cram
```

**Comment by rbuels** [comment]
Those are some very nice compression ratios and runtimes!

Regarding renaming: do you mean that you want to publish `@jkbonfield/htscodecs` on NPM under a different package name? Can't merge this until the package dependencies are all settled.

**Comment by jkbonfield** [comment]
I forgot about this comment:

> Thanks so much for this James. Would it be possible for you to add some automated tests for these new features? One good way to start that might be to make a file called `test/cram3.1.test.js` that runs some decompressions of small CRAM 3.1 test files that live in `test/data/`

We could add cram 3.1 equivalents to the test/data/*.cram.  It may need a bit of coercion on such small files to use the new codecs, but that's doable.  The actual javascript for htscodecs already has its own test data specifically for each codec, so for this project it's more of an integration test I guess.

Regarding the naming of the npm module, I'm not at all familier with npm policy and whether a subsequent rename would bring down the entire npm edifice.  We have a GA4GH file formats conference call tomorrow though so I've added official migration of htscodecs to the organisation onto the agenda.

For what it's worth, I'm not a JavaScript developer really and don't care much about how this is done.  I'd even be happy with the files just copy/paste into this repo if that's what you prefer, but probably you'd rather have someone else maintaining them.  I will add though that the intention is a bit different.  Here code is meant to be efficient and fit for purpose.  The htscodecs/javascript subdir is meant to be an easy to read demonstration of the spec, meaning it doesn't optimise much for efficiency and there is minimal room for error checking as it's just extra verbage when it comes to specification documents.  Hence long term it may not even be appropriate to label it as an official package.  Thoughts?

**Comment by rbuels** [comment]
Re: htscodecs implementations: for long-term maintainability it might be easiest for you or me to just integrate and "productionize" the htscodecs javascript files in this repository then, if you're not interested in maintaining an npm package. Maintaining them along with the rest of cram-js wouldn't be much of an additional burden there.

Sounds like a good plan for taking this to the finish line might be:

* make sure the CRAM 3.1 spec and the basic functionality of the htscodecs JS implementations are final and stable
* copy/port the JS implementations and their tests from htscodecs into this repo as part of this pull request
* add a few automated tests to test the integration and guard against regressions

What do you think of that plan?

**Comment by jkbonfield** [comment]
Works for me.  I'll let you know what we conclude in our GA4GH call tomorrow.

**Comment by rbuels** [comment]
@jkbonfield what ended up happening with the GA4GH call? Is the CRAM 3.1 spec finalized?

**Comment by jkbonfield** [comment]
Sorry for not getting back to you on this.

Sadly not because GA4GH requires two implementations (both read and write) and no one is doing a second one.  It's stalled it for some time.  The C and JavaScript ones actually count as two implementations, but as we don't have read/write support in JS yet it's not sufficiently complete.

It's possible GATK may be gaining support at some time, but not likely to the end of the year if not next year.

**Comment by cmdcolin** [comment]
Is it worth it to merge any parts of this PR? The bzip2 part is pretty cool, that is plain cram 3 right?

**Comment by cmdcolin** [comment]
Seems like this could probably be acceptable to merge in whole even since it enables future parsing of cram without changing anything about our backwards compatibility

**Comment by jkbonfield** [comment]
Bzip2 is plain cram3 (I think cram2 actually).

I'd still think about this comment from above though:

> For what it's worth, I'm not a JavaScript developer really and don't care much about how this is done. I'd even be happy with the files just copy/paste into this repo if that's what you prefer, but probably you'd rather have someone else maintaining them. I will add though that the intention is a bit different. Here (cram-js) code is meant to be efficient and fit for purpose. The htscodecs/javascript subdir is meant to be an easy to read demonstration of the spec, meaning it doesn't optimise much for efficiency and there is minimal room for error checking as it's just extra verbage when it comes to specification documents. Hence long term it may not even be appropriate to label it as an official package. Thoughts?

My own implementation of these codecs is more a demonstration for the specification, so they don't contain huge amounts of error checking (who am I kidding?  almost none) and I'm not really a Javascript developer so don't really understand npm and whether it's wise to just shove the stuff up there without long-term thinking.

We did discuss ownership of code - whether it should be placed under a GA4GH npm user for example.  However I think the general view is as the code here is more example rather than polished and supported, it may be best if it's simply sucked in to the only CRAM javascript code repository we know of... this :)

So I'll leave it up to you how you wish to proceed, but obviously there's no immediate rush.  That does mean if you want bzip2 of course you can just cherry pick in that little bit.

**Comment by cmdcolin** [comment]
Just thought I'd note that I used some of your code for generating a CIGAR/seq from this branch for use in our jbrowse application. Hope that's ok :) It would be nice to get something like this added to cram-js repo still too

**Comment by cmdcolin** [comment]
3.1 now published as a paper https://twitter.com/BonfieldJames/status/1481950839602302980

may be worth revisiting this sometime here...pre-emptively beat finding a non-working file in the wild

**Comment by cmdcolin** [comment]
I refreshed this PR in https://github.com/GMOD/cram-js/pull/102  due to some whitespace changes in the master branch making this pr a bit difficult to merge directly

**Comment by cmdcolin** [comment]
closed by #102 :)

**Pull request closed without merging**